feat!: remove hardcopy

lewis6991 opened the pull request:
Revive: #15235

lewis6991 committed:
feat!: remove hardcopy

Co-authored-by: Justin M. Keyes <justinkz@gmail.com>

zeertzjq commented:
- Make `has('printer')` return 0
- Make `has('postscript')` return 0
- Make `vim-patch.sh` remove the removed the files

zeertzjq commented on src/nvim/testdir/test_modeline.vim:
```diff
@@ -234,9 +234,7 @@ func Test_modeline_fails_always()
   call s:modeline_fails('omnifunc', 'omnifunc=Something()', 'E520:')
   call s:modeline_fails('operatorfunc', 'operatorfunc=Something()', 'E520:')
   call s:modeline_fails('perldll', 'perldll=Something()', 'E520:')
-  call s:modeline_fails('printdevice', 'printdevice=Something()', 'E520:')
   call s:modeline_fails('patchexpr', 'patchexpr=Something()', 'E520:')
-  call s:modeline_fails('printexpr', 'printexpr=Something()', 'E520:')
```
revert

zeertzjq commented:
I don't think your last commit remove translations properly. You didn't remove the comment above them, and for some fuzzy `msgstr` you didn't even remove `msgid`.

justinmk commented:
Updating my comment from last time... https://github.com/neovim/neovim/pull/15235#issuecomment-899482316 : 

> It's a niche feature, but has not been a noticeable burden to keep around. Except hardcopy.c does get flagged by static analysis, that's a cost.
> 
> If we can ship or point to a plugin that addresses the basic use-case, then we should rip out hardcopy. There is no reason for a bunch of builtin options for this. 
> 
> The main features needed AFAICT, are:
> 
> 1. get a full "rendered" document
>     - the `:TOhtml` plugin already demonstrates the mechanics of this (potentially improved by treesitter)
> 2. send it to the system printer
>     - does the Nvim API need a `hardcopy` event that UIs could handle?

Perhaps we could alias `:hardcopy` to TOHtml or something like that. We should try to have a reasonable answer to address the `:hardcopy` use-case, some sort of minimal plugin that outputs HTML or PDF, etc.

lewis6991 commented:
> I don't think your last commit removed translations properly. You didn't remove the comment above them, and for some fuzzy `msgstr` you didn't even remove `msgid`.

I removed them via `sed`. Cleaned them up now. Need to review them by hand.

lewis6991 commented:
> Perhaps we could alias :hardcopy to TOHtml or something like that. We should try to have a reasonable answer to address the :hardcopy use-case, some sort of minimal plugin that outputs HTML or PDF, etc.

The main issue is that `:hardcopy` behaves completely differently on Windows so I don't know what reasonable alias we could add.

For reference, the default value for `printexpr` (on unix) is:

```vim
    system(['lpr']
           + (empty(&printdevice)?[]:['-P', &printdevice])
           + [v:fname_in])
    .. delete(v:fname_in)
    + v:shell_error
```

So we could do something like:

```lua
    vim.fn.system({'lpr', vim.v.fname_in})
    vim.fn.delete(vim.v.fname_in)
    if vim.v.shell_error then
       error(vim.v.shell_error)
    end
```

Also we can't easily alias `:hardcopy` since it begins with a lowercase char, unless we implement the alias directly in C.

The use case is so niche I'm not sure if it would be worth it instead of just suggesting to use `:TOhtml` and a browser.

dundargoc commented:
Aliasing commands like that will age badly I think. Doing that will implicitly encode historical details that is gonna be annoying to explain. "Oh, in vim hardcopy prints out a file. In neovim, hardcopy is aliased to tohtml because this and this and that...". Finding information about hardcopy online is gonna be annoying as well, because confusions of the nature "do you mean vim hardcopy or neovim hardcopy?" will show up.

clason commented:
Agreed; better to provide a new (plugin) command `:Print`.

zeertzjq commented:
`:Print` may also cause confusion because it is equivalent to `:print` in Vim.

clason commented:
> Perhaps we could alias :hardcopy to TOHtml or something like that. We should try to have a reasonable answer to address the :hardcopy use-case, some sort of minimal plugin that outputs HTML or PDF, etc.

I think everyone is agreed that we want a rewrite of the `TOHtml` plugin that also handles extmarks, and that that would address the biggest chunk of use cases here.

The big question is: does that need to happen **before** the excision?  Arguably, `:hardcopy` is already broken right now as it doesn't work with treesitter or LSP semantic tokens (or any kind of extmark); seeing just how much code and docs is affected here further tilts my opinion in favor of removal.

(Whether and how to wrap `:TOHtml` into a cross-platform one-stop-print-shop is another question that can be tackled after these two steps.)

bfredl commented:
> The big question is: does that need to happen before the excision? Arguably, :hardcopy is already broken right now as it doesn't work with treesitter or LSP semantic tokens (or any kind of extmark); seeing just how much code and docs is affected here further tilts my opinion in favor of removal.

Let's nix it already. it is one legacy file less to be affected when doing general code refactors/quality improvements.

zeertzjq commented:
What is the point of removing this from menu.vim and optwin.vim then?

clason commented:
The possibility of users hitting errors when leaving it in?

zeertzjq commented:
They are already guarded.

bfredl commented:
Then we can just as well remove them.

lewis6991 commented:
Let's get some reviews so we can make a decision. I've just had to resolve a conflict so that file is getting churn.

bfredl commented on src/nvim/statusline.c:
```diff
@@ -1507,9 +1507,6 @@ int build_stl_str_hl(win_T *wp, char *out, size_t outlen, char *fmt, char *opt_n
         str = buf_tmp;
       }
       break;
-    case STL_PAGENUM:
```
now using `%N` in a statusline becames a hard error. is this intentional? The alternative (what vim `-printer` does) is to change this to `num = 0`

lewis6991 replied:
I didn't put much thought into it. I guess it would be safer to just return 0, assuming anyone even uses this.

justinmk commented on runtime/doc/vim_diff.txt:
```diff
@@ -732,5 +739,10 @@ Cscope:
 
+Hardcopy:
+                                                                     *hardcopy*
+  `:hardcopy` has been removed in favour of using `:TOhtml` and printing via
+  an external process.
```
This is a case where passive voice is making things awkward.
```suggestion
  `:hardcopy` was removed. Instead, use `:TOhtml` and print the resulting HTML using a web browser or some other HTML viewer.
```

zeertzjq commented on runtime/doc/vim_diff.txt:
```diff
@@ -732,5 +739,9 @@ Cscope:
 
+Hardcopy:
+                                                                     *hardcopy*
+  `:hardcopy` was removed. Instead, use `:TOhtml` and print the resulting HTML using a web browser or some other HTML viewer.
```
This line is too long.

lervag commented:
Do I understand correctly that the current fallback for those of us who used this feature is to now instead use `:TOhtml` and a browser? I personally don't like this, as it adds an extra step, but I can see the value of simplifying and I can agree with the idea that this feature would be better implemented as a plugin.

lewis6991 commented:
That's exactly the case.

- Most users didn't even know this existed
- of those that did,[ very few ever used it](https://twitter.com/Neovim/status/1433525597821562889)
- It didn't behave consistently across platforms
- It doesn't work with extmarks/treesiter highlights (just `:syntax`)

so the scales tipped pretty heavily to removing it.

If anyone creates a plugin for printer support, we would happily endorse it (providing it's good and works).

justinmk commented:
> now instead use `:TOhtml` and a browser?

I'm in favor of adding some builtin sugar that saves steps here. E.g. we could ship a `:Hardcopy` command that does the following:

1. generates HTML
2. opens the HTML in system web browser

Then you just hit ctrl-p to print, is that good enough? Step (2) above is the annoying part--finding and opening the file.

lervag commented:
> That's exactly the case. … so the scales tipped pretty heavily to removing it.

No objections from me, except the remark that this means, as one of the few users, that I will need to change my workflow. No big deal - I don't print things often anyway.

> > now instead use `:TOhtml` and a browser?
> 
> I'm in favor of adding some builtin sugar that saves steps here. E.g. we could ship a `:Hardcopy` command … , is that good enough?

Yes, for me it seems like a convenient fallback solution that I would be happy to use.

lf- commented:
I also use this. I'm not mad that it's been removed given the issues with tree sitter though; I have a hack script to turn it off temporarily, change colour scheme, etc to print, but this is a hack. The correct solution is some plugin rendering html then using a (possibly headless) web browser to generate a PDF. 

For my limited use cases I'll just use vim for it or something like https://github.com/matze/tree-painter

jmtd commented:
Am I right that the plan was to write a plugin to replace this, but that doesn’t exist yet, and the feature has been deleted anyway?

(Yes, I use it; no, I didn’t see nor reply to a random twitter poll)

lewis6991 commented:
> Am I right that the plan was to write a plugin to replace this, but that doesn’t exist yet, and the feature has been deleted anyway?

Yes.

jmtd commented:
Thank you for clarifying. Honestly I can most likely live without hardcopy these days. But it does make me wonder what other features I rely on might suddenly disappear. I need to rethink getting invested.

clason commented:
If you rely on a feature, you can always keep using the last version that still has it.

justinmk commented:
The path forward for hardcopy is https://github.com/neovim/neovim/issues/21629

justinmk commented:
https://github.com/neovim/neovim/pull/27097 adds treesitter (and more) support to `:TOhtml`, which opens the door for  `:Hardcopy`.
